ir/src/passes/inlining.rs

```diff
@@ -0,0 +1,211 @@
+use std::{collections::BTreeMap, ops::ControlFlow};
+
+use air_pass::Pass;
+//use miden_diagnostics::DiagnosticsHandler;
+
+use crate::{MirGraph, NodeIndex, Operation};
+
+//pub struct Inlining<'a> {
+//     #[allow(unused)]
+//     diagnostics: &'a DiagnosticsHandler,
+//}
+
+pub struct Inlining {}
+
+//impl<'p> Pass for Inlining<'p> {}
+impl Pass for Inlining {
+    type Input<'a> = MirGraph;
+    type Output<'a> = MirGraph;
+    type Error = ();
+
+    fn run<'a>(&mut self, mut ir: Self::Input<'a>) -> Result<Self::Output<'a>, Self::Error> {
+        match self.run_visitor(&mut ir) {
+            ControlFlow::Continue(()) => Ok(ir),
+            ControlFlow::Break(err) => Err(err),
+        }
+    }
+}
+
+// impl<'a> Inlining<'a> {
+//     pub fn new(diagnostics: &'a DiagnosticsHandler) -> Self {
+//         Self { diagnostics }
+//         Self {}
+//     }
+// }
+impl Inlining {
+    pub fn new() -> Self {
+        Self {}
+    }
+    //TODO MIR: Implement inlining pass on MIR
+    // 1. Understand the basics of the previous inlining process
+    // 2. Remove what is done during lowering from AST to MIR (unroll, ...)
+    // 3. Check how it translates to the MIR structure
+    fn run_visitor(&mut self, ir: &mut MirGraph) -> ControlFlow<()> {
+        inline_all(ir);
+        ControlFlow::Continue(())
+    }
+}
+
+fn inline_all(ir: &mut MirGraph) {
```

Comment:
One of the things I was hoping we could achieve with the new IR was avoiding unbounded recursion to visit the graph and perform analysis/transformation. Writing it in recursive style for the initial implementation is fine, but we should try and refactor it into iterative style so that the pass can handle arbitrarily deep graphs without panicking.

Reply:
The naming might be causing confusion here, there is no unbounded recursion, just a while loop calling inline_all -> inline_call -> inline_body -> inline_op. I'll try and simplify this once the proposed Node design is implemented with back-references via Op.owner, and will rename the functions to differentiate more easily between inline_all and inline_call.

Reply:
Ok, I was under the impression that `inline_call` transitively visits all inlineable items reachable from the root node being visited, but if I understand your description correctly, it is only visiting the immediate child of the `Call` node, i.e. it isn't also inlining calls in the callee, just the body of the callee without further inlining. Is that correct?

If so, then yeah we're good - I think I assumed the inliner was recursing into children, which is what prompted my comment.